<a href="https://colab.research.google.com/github/bugrakaann/SuspiciousActionRecognition/blob/master/x3dxs_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')


!pip install pytorchvideo torchvision

Mounted at /content/drive
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.7/132.7 kB 6.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.2/50.2 kB 3.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.2/42.2 kB 3.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 138.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 94.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 48.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 46.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/1

In [2]:
import os
import pandas as pd
import cv2
from sklearn.model_selection import train_test_split
from torch.utils.data import Dataset, DataLoader
import numpy as np

In [3]:
import gdown
import zipfile

file_id = '1TNnf9zzoreiF7qaWZFjRXEDejThIbXJe'
url = f"https://drive.google.com/uc?export=download&confirm=t&id={file_id}"

# Dosyayı indir
output_path = '/content/sample_data/dataset.zip'
gdown.download(url, output_path, quiet=False)

# Dosyayı unzip yap
with zipfile.ZipFile(output_path, 'r') as zip_ref:
    zip_ref.extractall('/content/sample_data')

print("Dosya başarıyla indirildi ve çıkarıldı.")

Downloading...
From: https://drive.google.com/uc?export=download&confirm=t&id=1TNnf9zzoreiF7qaWZFjRXEDejThIbXJe
To: /content/sample_data/dataset.zip
100%|██████████| 10.3G/10.3G [02:09<00:00, 79.3MB/s]


Dosya başarıyla indirildi ve çıkarıldı.


In [4]:
import os
import pandas as pd

# 🔥 Ana dataset klasörü
dataset_path = "/content/sample_data/dataset/Videos/Videos/"

# 📌 Normal ve Abnormal klasörlerini al
normal_path = os.path.normpath(os.path.join(dataset_path, "normal"))
abnormal_path = os.path.normpath(os.path.join(dataset_path, "abnormal"))

# 📌 Tüm abnormal sınıfları al
abnormal_classes = [folder for folder in os.listdir(abnormal_path) if os.path.isdir(os.path.join(abnormal_path, folder))]
class_to_index = {cls: idx for idx, cls in enumerate(abnormal_classes, start=1)}  # 1, 2, 3... (Abnormal sınıflar)

# 📌 DataFrame oluşturmak için liste
data = []

# ✅ Normal videolar (Binary = 0, Multi = -1)
for subfolder in os.listdir(normal_path):
    subfolder_path = os.path.normpath(os.path.join(normal_path, subfolder))

    if os.path.isdir(subfolder_path):  # Eğer bir klasörse
        videos = [
            os.path.normpath(os.path.join(subfolder_path, v)).replace("\\", "/")  # Windows uyumu için düzeltilmiş
            for v in os.listdir(subfolder_path) if v.endswith(('.mp4', '.avi', '.mov'))
        ]
        for video in videos:
            data.append((video, "normal", 0, -1))  # Binary label = 0, Multi label = -1 (yok)

# ✅ Anormal videolar (Binary = 1, Multi = class index)
for subfolder in abnormal_classes:
    subfolder_path = os.path.normpath(os.path.join(abnormal_path, subfolder))

    if os.path.isdir(subfolder_path):
        # Rekürsif olarak tüm alt klasörleri gezer
        for root, dirs, files in os.walk(subfolder_path):
            for file in files:
                if file.endswith(('.mp4', '.avi', '.mov')):
                    video_path = os.path.normpath(os.path.join(root, file)).replace("\\", "/")
                    # Burada 'subfolder' en üst düzey klasör adını temsil eder;
                    # eğer alt klasörlere göre farklı etiket atamak isterseniz, root veya dosya adını parse edebilirsiniz.
                    data.append((video_path, subfolder, 1, class_to_index.get(subfolder, -1)))

# 🔥 DataFrame oluştur
df = pd.DataFrame(data, columns=["video_path", "category", "binary_label", "multi_label"])


In [5]:
train_df, test_df = train_test_split(df, test_size=0.2,stratify=df["binary_label"],random_state=42)

print(f"Train set {len(train_df)} videos")
print(f"Test set {len(test_df)} videos")

Train set 2603 videos
Test set 651 videos


In [6]:
import tensorflow as tf

# GPU kullanımını kontrol et
physical_devices = tf.config.list_physical_devices('GPU')
if physical_devices:
    print("GPU bulunuyor:", physical_devices)
else:
    print("GPU bulunamadı.")


GPU bulunamadı.


In [7]:
!pip install tensorflow==2.15.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 475.3/475.3 MB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 68.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 57.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 109.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.6/294.6 kB 22.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 125.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 442.0/442.0 kB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.4/78.4 kB 6.1 MB/s eta 0:00:00
  Attempting uninstall: wrapt
    Found existing installation: wrapt 1.17.2
    Uninstalling wrapt-1.17.2:
      Successfully uninstalled wrapt-1.17.2
  Attempting uninstall: protobuf
    Found existing installation: protobuf 5.29.4
    Uninstalling protobuf-5.29.4:
      Successfully un

In [ ]:
import os
import tensorflow as tf
import cv2
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.layers import Input, Conv2D, MaxPooling2D, UpSampling2D, Dense, Reshape, LSTM
from tensorflow.keras.models import Model
from tensorflow.keras.layers import ConvLSTM2D, BatchNormalization
from tqdm import tqdm
import pandas as pd

# Tensorflow session'u temizle (önceki işlemlerden kalanları sıfırla)
tf.keras.backend.clear_session()

# GPU kullanımını devre dışı bırakma
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, False)  # GPU'yu devre dışı bırakıyoruz
        print("GPU devre dışı bırakıldı.")
    except RuntimeError as e:
        print(f"GPU ayarlanırken hata oluştu: {e}")
else:
    print("GPU bulunamadı, CPU kullanılacak.")

# Tensorflow'un CUDA ile derlenip derlenmediğini ve fiziksel GPU olup olmadığını göster
print("TensorFlow CUDA destekli mi?:", tf.test.is_built_with_cuda())
print("Fiziksel GPU var mı?:", tf.config.list_physical_devices('GPU'))


def build_autoencoder(input_shape):
    # Input shape: (height, width, channels)
    input_img = Input(shape=input_shape, name="input_layer")

    # Encoder kısmı
    x = Conv2D(64, (3, 3), activation='relu', padding='same')(input_img)
    x = BatchNormalization()(x)
    x = MaxPooling2D((2, 2), padding='same')(x)

    x = Conv2D(32, (3, 3), activation='relu', padding='same')(x)
    x = BatchNormalization()(x)
    encoded = MaxPooling2D((2, 2), padding='same')(x)

    # Temporal özellikler için 1x1 konvolüsyon kullanın
    temporal_features = Conv2D(16, (1, 1), activation='relu')(encoded)

    # Decoder kısmı
    x = Conv2D(32, (3, 3), activation='relu', padding='same')(temporal_features)
    x = BatchNormalization()(x)
    x = UpSampling2D((2, 2))(x)

    x = Conv2D(64, (3, 3), activation='relu', padding='same')(x)
    x = BatchNormalization()(x)
    x = UpSampling2D((2, 2))(x)

    # Çıkış - orijinal görüntüyü yeniden oluştur
    decoded = Conv2D(input_shape[2], (3, 3), activation='sigmoid', padding='same')(x)

    # Model tanımı
    autoencoder = Model(input_img, decoded)
    autoencoder.compile(optimizer='adam', loss='mse')

    return autoencoder


# Frame yükleme fonksiyonu - GPU için optimize edilmiş
def load_frames_from_videos(folder, target_size, frame_interval=30):
    frames = []
    valid_extensions = ('.mp4', '.avi', '.mkv')

    for root, _, files in os.walk(folder):
        for filename in sorted(files):
            if not filename.lower().endswith(valid_extensions):
                continue

            video_path = os.path.join(root, filename)
            cap = cv2.VideoCapture(video_path)
            frame_idx = 0
            total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))

            with tqdm(total=total_frames, desc=f"İşleniyor: {filename}", leave=False) as pbar:
                while cap.isOpened():
                    ret, frame = cap.read()
                    if not ret:
                        break

                    if frame_idx % frame_interval == 0:
                        frame_resized = cv2.resize(frame, target_size)
                        frame_rgb = cv2.cvtColor(frame_resized, cv2.COLOR_BGR2RGB)
                        frames.append(frame_rgb)

                    frame_idx += 1
                    pbar.update(1)

            cap.release()
            tqdm.write(f"{filename} dosyasından {frame_idx // frame_interval} kare çekildi.")

    if len(frames) == 0:
        raise ValueError(f"'{folder}' klasöründe geçerli video dosyası bulunamadı.")

    # GPU için optimize edilmiş veri hazırlama
    frames = np.stack(frames, axis=0).astype('float32') / 255.0
    return frames


# Eğitim fonksiyonu
def train_autoencoder(normal_videos_folder, target_size=(128, 128), batch_size=32, epochs=100):
    # Veri yükleme
    print("Normal videolardan frame'ler yükleniyor...")
    normal_frames = load_frames_from_videos(normal_videos_folder, target_size)
    print(f"Toplam {len(normal_frames)} frame yüklendi.")

    # Model oluşturma
    input_shape = (target_size[1], target_size[0], 3)
    autoencoder = build_autoencoder(input_shape)
    autoencoder.summary()

    # ModelCheckpoint Callback
    checkpoint_dir = '/content/drive/MyDrive/modelepochsaves/'
    if not os.path.exists(checkpoint_dir):
        os.makedirs(checkpoint_dir)

    callbacks = [
        tf.keras.callbacks.EarlyStopping(
            monitor='val_loss',
            patience=5,
            restore_best_weights=True
        ),
        tf.keras.callbacks.ReduceLROnPlateau(
            monitor='val_loss',
            factor=0.2,
            patience=3,
            min_lr=1e-6
        ),
        tf.keras.callbacks.LambdaCallback(
            on_epoch_end=lambda epoch, logs: autoencoder.save(
                f'{checkpoint_dir}autoencoder_epoch_{epoch+1}'  # Epoch sayısı eklenmiş dosya adı
            )
        )
    ]

    # Eğitim
    print("Autoencoder eğitimi başlıyor...")
    history = autoencoder.fit(
        normal_frames, normal_frames,
        validation_split=0.1,
        epochs=epochs,
        batch_size=batch_size,
        shuffle=True,
        callbacks=callbacks
    )

    return autoencoder, history


# Eğitimi başlat
normal_videos_folder = '/content/sample_data/dataset/Videos/Videos/normal'
autoencoder, history = train_autoencoder(normal_videos_folder)

# Eğitim sonuçlarını görselleştir
plt.figure(figsize=(12, 4))
plt.subplot(1, 2, 1)
plt.plot(history.history['loss'], label='Training Loss')
plt.plot(history.history['val_loss'], label='Validation Loss')
plt.title('Model Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()

# Modeli kaydet
autoencoder.save(f'/content/drive/MyDrive/modelepochsaves/autoencoder_final_{datetime.datetime.now().strftime("%Y%m%d-%H%M%S")}')
print("Model kaydedildi: autoencoder_final")


GPU bulunamadı, CPU kullanılacak.
TensorFlow CUDA destekli mi?: True
Fiziksel GPU var mı?: []
Normal videolardan frame'ler yükleniyor...


n001_converted.avi dosyasından 5 kare çekildi.


n002_converted.avi dosyasından 5 kare çekildi.


n003_converted.avi dosyasından 4 kare çekildi.


n004_converted.avi dosyasından 5 kare çekildi.


n005_converted.avi dosyasından 5 kare çekildi.


n006_converted.avi dosyasından 5 kare çekildi.


n007_converted.avi dosyasından 4 kare çekildi.


n008_converted.avi dosyasından 4 kare çekildi.


n009_converted.avi dosyasından 4 kare çekildi.


n010_converted.avi dosyasından 4 kare çekildi.


n011_converted.avi dosyasından 4 kare çekildi.


n012_converted.avi dosyasından 5 kare çekildi.


n013_converted.avi dosyasından 4 kare çekildi.


n014_converted.avi dosyasından 4 kare çekildi.


n015_converted.avi dosyasından 4 kare çekildi.


n016_converted.avi dosyasından 4 kare çekildi.


n017_converted.avi dosyasından 4 kare çekildi.


n018_converted.avi dosyasından 4 kare çekildi.


n019_converted.avi dosyasından 4 kare çekildi.


n020_converted.avi dosyasından 4 kare çekildi.


n021_converted.avi dosyasından 4 kare çekildi.


n022_converted.avi dosyasından 4 kare çekildi.


n023_converted.avi dosyasından 4 kare çekildi.


n024_converted.avi dosyasından 4 kare çekildi.


n025_converted.avi dosyasından 3 kare çekildi.


n026_converted.avi dosyasından 3 kare çekildi.


n027_converted.avi dosyasından 3 kare çekildi.


n028_converted.avi dosyasından 4 kare çekildi.


n029_converted.avi dosyasından 5 kare çekildi.


n030_converted.avi dosyasından 5 kare çekildi.


n031_converted.avi dosyasından 4 kare çekildi.


n032_converted.avi dosyasından 4 kare çekildi.


n033_converted.avi dosyasından 5 kare çekildi.


n034_converted.avi dosyasından 5 kare çekildi.


n035_converted.avi dosyasından 6 kare çekildi.


n036_converted.avi dosyasından 4 kare çekildi.


n037_converted.avi dosyasından 5 kare çekildi.


n038_converted.avi dosyasından 5 kare çekildi.


n039_converted.avi dosyasından 4 kare çekildi.


n040_converted.avi dosyasından 5 kare çekildi.


n041_converted.avi dosyasından 5 kare çekildi.


n042_converted.avi dosyasından 5 kare çekildi.


n043_converted.avi dosyasından 5 kare çekildi.


n044_converted.avi dosyasından 4 kare çekildi.


n045_converted.avi dosyasından 5 kare çekildi.


n046_converted.avi dosyasından 4 kare çekildi.


n047_converted.avi dosyasından 5 kare çekildi.


n048_converted.avi dosyasından 5 kare çekildi.


n049_converted.avi dosyasından 5 kare çekildi.


n050_converted.avi dosyasından 5 kare çekildi.


n051_converted.avi dosyasından 4 kare çekildi.


n052_converted.avi dosyasından 4 kare çekildi.


n053_converted.avi dosyasından 5 kare çekildi.


n054_converted.avi dosyasından 4 kare çekildi.


n055_converted.avi dosyasından 4 kare çekildi.


n056_converted.avi dosyasından 3 kare çekildi.


n057_converted.avi dosyasından 4 kare çekildi.


n058_converted.avi dosyasından 4 kare çekildi.


n059_converted.avi dosyasından 4 kare çekildi.


n060_converted.avi dosyasından 4 kare çekildi.


n061_converted.avi dosyasından 4 kare çekildi.


n062_converted.avi dosyasından 4 kare çekildi.


n063_converted.avi dosyasından 3 kare çekildi.


n064_converted.avi dosyasından 5 kare çekildi.


n065_converted.avi dosyasından 5 kare çekildi.


n066_converted.avi dosyasından 5 kare çekildi.


n067_converted.avi dosyasından 6 kare çekildi.


n068_converted.avi dosyasından 5 kare çekildi.


n069_converted.avi dosyasından 4 kare çekildi.


n070_converted.avi dosyasından 4 kare çekildi.


n071_converted.avi dosyasından 4 kare çekildi.


n072_converted.avi dosyasından 4 kare çekildi.


n073_converted.avi dosyasından 4 kare çekildi.


n074_converted.avi dosyasından 4 kare çekildi.


n075_converted.avi dosyasından 4 kare çekildi.


n076_converted.avi dosyasından 3 kare çekildi.


n077_converted.avi dosyasından 4 kare çekildi.


n078_converted.avi dosyasından 4 kare çekildi.


n079_converted.avi dosyasından 4 kare çekildi.


n080_converted.avi dosyasından 4 kare çekildi.


n081_converted.avi dosyasından 3 kare çekildi.


n082_converted.avi dosyasından 4 kare çekildi.


n083_converted.avi dosyasından 4 kare çekildi.


n084_converted.avi dosyasından 3 kare çekildi.


n085_converted.avi dosyasından 3 kare çekildi.


n086_converted.avi dosyasından 5 kare çekildi.


n087_converted.avi dosyasından 4 kare çekildi.


n088_converted.avi dosyasından 4 kare çekildi.


n089_converted.avi dosyasından 5 kare çekildi.


n090_converted.avi dosyasından 4 kare çekildi.


n091_converted.avi dosyasından 4 kare çekildi.


n092_converted.avi dosyasından 4 kare çekildi.


n093_converted.avi dosyasından 4 kare çekildi.


n094_converted.avi dosyasından 3 kare çekildi.


n095_converted.avi dosyasından 3 kare çekildi.


n096_converted.avi dosyasından 3 kare çekildi.


n097_converted.avi dosyasından 4 kare çekildi.


n098_converted.avi dosyasından 3 kare çekildi.


n099_converted.avi dosyasından 5 kare çekildi.


n100_converted.avi dosyasından 5 kare çekildi.


n101_converted.avi dosyasından 5 kare çekildi.


n102_converted.avi dosyasından 4 kare çekildi.


n103_converted.avi dosyasından 4 kare çekildi.


n104_converted.avi dosyasından 4 kare çekildi.


n105_converted.avi dosyasından 4 kare çekildi.


n106_converted.avi dosyasından 4 kare çekildi.


n107_converted.avi dosyasından 4 kare çekildi.


n108_converted.avi dosyasından 5 kare çekildi.


n109_converted.avi dosyasından 4 kare çekildi.


n110_converted.avi dosyasından 5 kare çekildi.


n111_converted.avi dosyasından 3 kare çekildi.


n112_converted.avi dosyasından 3 kare çekildi.


n113_converted.avi dosyasından 9 kare çekildi.


n114_converted.avi dosyasından 8 kare çekildi.


n115_converted.avi dosyasından 8 kare çekildi.


n116_converted.avi dosyasından 4 kare çekildi.


n117_converted.avi dosyasından 3 kare çekildi.


n118_converted.avi dosyasından 4 kare çekildi.


n119_converted.avi dosyasından 4 kare çekildi.


n120_converted.avi dosyasından 3 kare çekildi.


n121_converted.avi dosyasından 4 kare çekildi.


n122_converted.avi dosyasından 3 kare çekildi.


n123_converted.avi dosyasından 4 kare çekildi.


n124_converted.avi dosyasından 3 kare çekildi.


n125_converted.avi dosyasından 4 kare çekildi.


n126_converted.avi dosyasından 4 kare çekildi.


n127_converted.avi dosyasından 4 kare çekildi.


n128_converted.avi dosyasından 4 kare çekildi.


n129_converted.avi dosyasından 4 kare çekildi.


n130_converted.avi dosyasından 4 kare çekildi.


n131_converted.avi dosyasından 5 kare çekildi.


n132_converted.avi dosyasından 5 kare çekildi.


n133_converted.avi dosyasından 4 kare çekildi.


n134_converted.avi dosyasından 5 kare çekildi.


n135_converted.avi dosyasından 4 kare çekildi.


n136_converted.avi dosyasından 4 kare çekildi.


n137_converted.avi dosyasından 3 kare çekildi.


n138_converted.avi dosyasından 4 kare çekildi.


n139_converted.avi dosyasından 3 kare çekildi.


n140_converted.avi dosyasından 4 kare çekildi.


n141_converted.avi dosyasından 4 kare çekildi.


n142_converted.avi dosyasından 4 kare çekildi.


n143_converted.avi dosyasından 1 kare çekildi.


n144_converted.avi dosyasından 1 kare çekildi.


n145_converted.avi dosyasından 0 kare çekildi.


n146_converted.avi dosyasından 1 kare çekildi.


n147_converted.avi dosyasından 1 kare çekildi.


n148_converted.avi dosyasından 0 kare çekildi.


n149_converted.avi dosyasından 0 kare çekildi.


n150_converted.avi dosyasından 1 kare çekildi.


n151_converted.avi dosyasından 0 kare çekildi.


n152_converted.avi dosyasından 0 kare çekildi.


n153_converted.avi dosyasından 1 kare çekildi.


n154_converted.avi dosyasından 1 kare çekildi.


n155_converted.avi dosyasından 1 kare çekildi.


n156_converted.avi dosyasından 4 kare çekildi.


n157_converted.avi dosyasından 3 kare çekildi.


n158_converted.avi dosyasından 3 kare çekildi.


n159_converted.avi dosyasından 3 kare çekildi.


n160_converted.avi dosyasından 3 kare çekildi.


n161_converted.avi dosyasından 5 kare çekildi.


n162_converted.avi dosyasından 3 kare çekildi.


n163_converted.avi dosyasından 3 kare çekildi.


n164_converted.avi dosyasından 3 kare çekildi.


n165_converted.avi dosyasından 4 kare çekildi.


n166_converted.avi dosyasından 2 kare çekildi.


n167_converted.avi dosyasından 1 kare çekildi.


n168_converted.avi dosyasından 1 kare çekildi.


n169_converted.avi dosyasından 1 kare çekildi.


n170_converted.avi dosyasından 1 kare çekildi.


n171_converted.avi dosyasından 1 kare çekildi.


n172_converted.avi dosyasından 0 kare çekildi.


n173_converted.avi dosyasından 0 kare çekildi.


n174_converted.avi dosyasından 1 kare çekildi.


n175_converted.avi dosyasından 1 kare çekildi.


n176_converted.avi dosyasından 0 kare çekildi.


n177_converted.avi dosyasından 1 kare çekildi.


n178_converted.avi dosyasından 1 kare çekildi.


n179_converted.avi dosyasından 1 kare çekildi.


n180_converted.avi dosyasından 0 kare çekildi.


n181_converted.avi dosyasından 0 kare çekildi.


n182_converted.avi dosyasından 0 kare çekildi.


n183_converted.avi dosyasından 1 kare çekildi.


n184_converted.avi dosyasından 0 kare çekildi.


n185_converted.avi dosyasından 1 kare çekildi.


n186_converted.avi dosyasından 1 kare çekildi.


n187_converted.avi dosyasından 1 kare çekildi.


n188_converted.avi dosyasından 1 kare çekildi.


n189_converted.avi dosyasından 1 kare çekildi.


n190_converted.avi dosyasından 4 kare çekildi.


n191_converted.avi dosyasından 4 kare çekildi.


n192_converted.avi dosyasından 4 kare çekildi.


n193_converted.avi dosyasından 4 kare çekildi.


n194_converted.avi dosyasından 3 kare çekildi.


n195_converted.avi dosyasından 4 kare çekildi.


n196_converted.avi dosyasından 4 kare çekildi.


n197_converted.avi dosyasından 4 kare çekildi.


n198_converted.avi dosyasından 4 kare çekildi.


n199_converted.avi dosyasından 4 kare çekildi.


n200_converted.avi dosyasından 4 kare çekildi.


t_n001_converted.avi dosyasından 4 kare çekildi.


t_n002_converted.avi dosyasından 4 kare çekildi.


t_n003_converted.avi dosyasından 3 kare çekildi.


t_n004_converted.avi dosyasından 4 kare çekildi.


t_n005_converted.avi dosyasından 3 kare çekildi.


t_n006_converted.avi dosyasından 4 kare çekildi.


t_n007_converted.avi dosyasından 3 kare çekildi.


t_n008_converted.avi dosyasından 3 kare çekildi.


t_n009_converted.avi dosyasından 4 kare çekildi.


t_n010_converted.avi dosyasından 4 kare çekildi.


t_n011_converted.avi dosyasından 4 kare çekildi.


t_n012_converted.avi dosyasından 4 kare çekildi.


t_n013_converted.avi dosyasından 4 kare çekildi.


t_n014_converted.avi dosyasından 4 kare çekildi.


t_n015_converted.avi dosyasından 4 kare çekildi.


t_n016_converted.avi dosyasından 2 kare çekildi.


t_n017_converted.avi dosyasından 7 kare çekildi.


t_n018_converted.avi dosyasından 4 kare çekildi.


t_n019_converted.avi dosyasından 3 kare çekildi.


t_n020_converted.avi dosyasından 3 kare çekildi.


t_n021_converted.avi dosyasından 4 kare çekildi.


t_n022_converted.avi dosyasından 4 kare çekildi.


t_n023_converted.avi dosyasından 3 kare çekildi.


t_n024_converted.avi dosyasından 4 kare çekildi.


t_n025_converted.avi dosyasından 4 kare çekildi.


t_n026_converted.avi dosyasından 4 kare çekildi.


t_n027_converted.avi dosyasından 3 kare çekildi.


t_n028_converted.avi dosyasından 4 kare çekildi.


t_n029_converted.avi dosyasından 4 kare çekildi.


t_n030_converted.avi dosyasından 3 kare çekildi.


t_n031_converted.avi dosyasından 3 kare çekildi.


t_n032_converted.avi dosyasından 4 kare çekildi.


t_n033_converted.avi dosyasından 4 kare çekildi.


t_n034_converted.avi dosyasından 4 kare çekildi.


t_n035_converted.avi dosyasından 3 kare çekildi.


t_n036_converted.avi dosyasından 3 kare çekildi.


t_n037_converted.avi dosyasından 1 kare çekildi.


t_n038_converted.avi dosyasından 4 kare çekildi.


t_n039_converted.avi dosyasından 3 kare çekildi.


t_n040_converted.avi dosyasından 4 kare çekildi.


t_n041_converted.avi dosyasından 4 kare çekildi.


t_n042_converted.avi dosyasından 4 kare çekildi.


t_n043_converted.avi dosyasından 4 kare çekildi.


t_n044_converted.avi dosyasından 4 kare çekildi.


t_n045_converted.avi dosyasından 4 kare çekildi.


t_n046_converted.avi dosyasından 4 kare çekildi.


UCFCRIME_Abuse004_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse004_stand_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse005_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse006_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse008_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse008_stand_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse014_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse018_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse028_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse030_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse031_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse038_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse038_stand_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse039_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse039_stand_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse040_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Abuse040_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse044_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse045_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse046_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest001_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest003_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest003_stand_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest007_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest007_stand_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest009_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest011_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest014_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest014_stand_2.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Arrest016_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest016_stand_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest020_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest021_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest026_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest028_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest032_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest041_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest043_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest044_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest046_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest048_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson001_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson015_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson022_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson026_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson027_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson035_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson041_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson042_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arson044_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Assault013_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault014_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault015_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault016_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Assault019_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault023_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault027_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault030_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Assault041_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Assault047_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault050_stand_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Burglary002_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary003_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary005_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary029_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary052_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary076_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Fighting002_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Fighting003_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Fighting027_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Fighting042_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Fighting050_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery001_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery003_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery004_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery006_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery007_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery009_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery016_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery020_stand_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Robbery026_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery027_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery038_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery042_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery048_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery050_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery054_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery059_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery064_stand_1.mp4 dosyasından 9 kare çekildi.


UCFCRIME_Robbery071_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery079_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery080_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery092_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery098_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery098_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery101_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery106_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery107_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery115_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Robbery117_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery124_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery125_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery132_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery133_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery137_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery139_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery140_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery148_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery148_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting009_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting011_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting013_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting014_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting014_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting018_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting019_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting019_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting029_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting031_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting032_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting033_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting039_stand_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Shooting040_stand_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Shooting040_stand_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting043_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting043_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting044_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting046_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting046_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting047_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting048_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting050_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting053_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting001_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting003_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting004_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting005_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting006_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting007_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting009_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting010_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting012_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting014_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting015_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting015_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting016_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting017_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting020_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting021_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting025_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting027_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting028_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting029_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting029_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting030_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting030_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting032_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting033_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting034_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting036_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting038_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting039_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting040_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting041_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting042_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting043_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting044_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting044_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting050_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting051_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting052_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting052_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting053_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting055_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Stealing004_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing011_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing014_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing017_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing019_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing025_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing026_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing036_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing055_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism006_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Vandalism006_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism007_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Vandalism008_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism008_stand_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism010_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism015_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism015_stand_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism015_stand_3.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism019_stand_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Vandalism034_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Vandalism035_stand_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Vandalism036_stand_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Vandalism038_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism042_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism049_stand_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Vandalism049_stand_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse007_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse007_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse007_walk_3.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse014_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse014_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse018_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse022_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse028_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse028_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse028_walk_3.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse029_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse030_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse030_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse031_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Abuse031_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Abuse034_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Abuse046_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest001_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest001_walk_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest002_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest004_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest005_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest007_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest009_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest014_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest016_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest016_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest016_walk_3.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest020_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arrest022_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest022_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest026_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest029_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest029_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest029_walk_3.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest033_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest035_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest041_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest043_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest044_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest044_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest046_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest047_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arrest048_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arrest049_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Arson001_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson001_walk_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson002_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arson006_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson007_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson007_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Arson014_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson015_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson017_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson017_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson020_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson027_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson028_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson034_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson044_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson047_walk_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Arson052_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Assault010_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault010_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault011_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault011_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault014_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault016_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault016_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault034_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault040_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Assault047_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary005_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Burglary016_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary020_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary025_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary031_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Burglary039_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Burglary040_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Burglary040_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Burglary055_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary062_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Burglary063_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary066_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary067_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Burglary073_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Burglary076_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Burglary092_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Fighting021_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Fighting027_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Fighting030_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Fighting031_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Fighting031_walk_2.mp4 dosyasından 7 kare çekildi.


UCFCRIME_Fighting041_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Fighting041_walk_2.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Fighting043_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery009_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery021_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery027_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery032_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Robbery037_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery042_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery052_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery059_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery061_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery062_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery063_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery067_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Robbery069_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery069_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery072_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery073_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery080_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery091_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery094_walk_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Robbery095_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery096_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery097_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery099_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery103_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery106_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery112_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery119_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery119_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery135_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery150_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting010_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting014_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting014_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting022_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting028_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting031_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting032_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting032_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting038_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting040_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting040_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shooting047_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shooting047_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting047_walk_3.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting054_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting001_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting001_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting003_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting005_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting005_walk_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting010_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting012_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting014_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting015_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting016_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting018_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting020_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting026_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting027_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting027_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting029_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting029_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting032_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting032_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting034_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting038_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting039_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting039_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting040_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting044_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting047_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting051_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Shoplifting052_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting052_walk_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting053_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting055_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting055_walk_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting055_walk_3.mp4 dosyasından 7 kare çekildi.


UCFCRIME_Stealing002_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing003_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing004_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing004_walk_2.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing007_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing009_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing011_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing012_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing015_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing016_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing019_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing025_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing026_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing035_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing047_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing048_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing052_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing054_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing059_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing062_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing065_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing067_walk_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing078_walk_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing083_walk_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing091_walk_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing101_walk_1.mp4 dosyasından 4 kare çekildi.


Normal_Videos_015_x264.mp4 dosyasından 16 kare çekildi.


Normal_Videos_050_x264.mp4 dosyasından 139 kare çekildi.


Normal_Videos_100_x264.mp4 dosyasından 20 kare çekildi.


Normal_Videos_129_x264.mp4 dosyasından 15 kare çekildi.


Normal_Videos_150_x264.mp4 dosyasından 28 kare çekildi.


Normal_Videos_246_x264.mp4 dosyasından 166 kare çekildi.


Normal_Videos_247_x264.mp4 dosyasından 273 kare çekildi.


Normal_Videos_248_x264.mp4 dosyasından 38 kare çekildi.


Normal_Videos_251_x264.mp4 dosyasından 13 kare çekildi.


Normal_Videos_289_x264.mp4 dosyasından 28 kare çekildi.


Normal_Videos_310_x264.mp4 dosyasından 83 kare çekildi.


Normal_Videos_312_x264.mp4 dosyasından 42 kare çekildi.


Normal_Videos_317_x264.mp4 dosyasından 30 kare çekildi.


Normal_Videos_345_x264.mp4 dosyasından 6 kare çekildi.


Normal_Videos_352_x264.mp4 dosyasından 180 kare çekildi.


Normal_Videos_360_x264.mp4 dosyasından 32 kare çekildi.


Normal_Videos_365_x264.mp4 dosyasından 220 kare çekildi.


Normal_Videos_401_x264.mp4 dosyasından 54 kare çekildi.


Normal_Videos_417_x264.mp4 dosyasından 35 kare çekildi.


Normal_Videos_439_x264.mp4 dosyasından 140 kare çekildi.


Normal_Videos_452_x264.mp4 dosyasından 14 kare çekildi.


Normal_Videos_453_x264.mp4 dosyasından 177 kare çekildi.


Normal_Videos_478_x264.mp4 dosyasından 150 kare çekildi.


Normal_Videos_576_x264.mp4 dosyasından 375 kare çekildi.


Normal_Videos_597_x264.mp4 dosyasından 74 kare çekildi.


Normal_Videos_603_x264.mp4 dosyasından 109 kare çekildi.


Normal_Videos_606_x264.mp4 dosyasından 41 kare çekildi.


Normal_Videos_621_x264.mp4 dosyasından 160 kare çekildi.


Normal_Videos_634_x264.mp4 dosyasından 448 kare çekildi.


Normal_Videos_641_x264.mp4 dosyasından 120 kare çekildi.


Normal_Videos_656_x264.mp4 dosyasından 60 kare çekildi.


Normal_Videos_696_x264.mp4 dosyasından 120 kare çekildi.


Normal_Videos_704_x264.mp4 dosyasından 56 kare çekildi.


Normal_Videos_722_x264.mp4 dosyasından 291 kare çekildi.


Normal_Videos_745_x264.mp4 dosyasından 10 kare çekildi.


Normal_Videos_758_x264.mp4 dosyasından 52 kare çekildi.


Normal_Videos_781_x264.mp4 dosyasından 132 kare çekildi.


Normal_Videos_798_x264.mp4 dosyasından 200 kare çekildi.


Normal_Videos_801_x264.mp4 dosyasından 91 kare çekildi.


Normal_Videos_828_x264.mp4 dosyasından 31 kare çekildi.


Normal_Videos_831_x264.mp4 dosyasından 14 kare çekildi.


Normal_Videos_877_x264.mp4 dosyasından 334 kare çekildi.


Normal_Videos_881_x264.mp4 dosyasından 7 kare çekildi.


Normal_Videos_885_x264.mp4 dosyasından 15 kare çekildi.


Normal_Videos_892_x264.mp4 dosyasından 59 kare çekildi.


Normal_Videos_905_x264.mp4 dosyasından 39 kare çekildi.


Normal_Videos_912_x264.mp4 dosyasından 24 kare çekildi.


Normal_Videos_913_x264.mp4 dosyasından 20 kare çekildi.


Normal_Videos_914_x264.mp4 dosyasından 29 kare çekildi.


Normal_Videos_929_x264.mp4 dosyasından 30 kare çekildi.


NTU_fight0027_sit_1.mp4 dosyasından 2 kare çekildi.


NTU_fight0030_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0042_sit_1.mp4 dosyasından 6 kare çekildi.


NTU_fight0054_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0054_sit_2.mp4 dosyasından 4 kare çekildi.


NTU_fight0067_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0097_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0098_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0098_sit_2.mp4 dosyasından 4 kare çekildi.


NTU_fight0115_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0121_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0122_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0133_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0146_sit_1.mp4 dosyasından 2 kare çekildi.


NTU_fight0162_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0188_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0200_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0223_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0226_sit_1.mp4 dosyasından 6 kare çekildi.


NTU_fight0228_sit_1.mp4 dosyasından 5 kare çekildi.


NTU_fight0242_sit_1.mp4 dosyasından 5 kare çekildi.


NTU_fight0244_sit_1.mp4 dosyasından 3 kare çekildi.


NTU_fight0246_sit_1.mp4 dosyasından 5 kare çekildi.


NTU_fight0321_sit_1.mp4 dosyasından 1 kare çekildi.


NTU_fight0599_sit_1.mp4 dosyasından 5 kare çekildi.


NTU_fight0990_sit_1.mp4 dosyasından 5 kare çekildi.


NTU_fight0994_sit_1.mp4 dosyasından 4 kare çekildi.


NTU_fight0995_sit_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Abuse040_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Arson016_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Arson029_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault004_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault009_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Assault010_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Assault010_sit_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Burglary064_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Burglary064_sit_2.mp4 dosyasından 8 kare çekildi.


UCFCRIME_Fighting046_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Fighting049_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Fighting049_sit_2.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery019_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery020_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery024_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery026_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery037_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery045_sit_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Robbery062_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery073_sit_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Robbery080_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery090_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Robbery093_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Robbery093_sit_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Robbery097_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery123_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery126_sit_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Robbery127_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Robbery144_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Robbery146_sit_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Shooting006_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting017_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shooting022_sit_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Shooting030_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting031_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shooting048_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting003_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting007_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting008_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting010_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting012_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting012_sit_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Shoplifting024_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting037_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Shoplifting038_sit_1.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Shoplifting043_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting044_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Shoplifting051_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing009_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing011_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing022_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing024_sit_1.mp4 dosyasından 1 kare çekildi.


UCFCRIME_Stealing024_sit_2.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Stealing025_sit_1.mp4 dosyasından 5 kare çekildi.


UCFCRIME_Stealing036_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing042_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing044_sit_1.mp4 dosyasından 6 kare çekildi.


UCFCRIME_Stealing057_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing057_sit_2.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing057_sit_3.mp4 dosyasından 2 kare çekildi.


UCFCRIME_Stealing083_sit_1.mp4 dosyasından 3 kare çekildi.


UCFCRIME_Stealing096_sit_1.mp4 dosyasından 4 kare çekildi.


UCFCRIME_Stealing104_sit_1.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_10.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_11.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_12.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_13.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_2.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_3.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_4.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_5.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_6.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_7.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_8.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV001_sit_9.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV002_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV002_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV002_sit_3.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV002_sit_4.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV004_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV005_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV008_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_1.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_3.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_4.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_5.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_6.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV010_sit_7.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV012_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV014_sit_1.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV014_sit_2.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV014_sit_3.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV014_sit_4.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV014_sit_5.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV015_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV015_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV015_sit_3.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV015_sit_4.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV015_sit_5.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_2.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_3.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_4.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_5.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_6.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_7.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV017_sit_8.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV019_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV020_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV021_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV022_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV022_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV022_sit_3.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV030_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV036_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV036_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV036_sit_3.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV042_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV042_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV042_sit_3.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV044_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV044_sit_2.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_2.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_3.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_4.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_5.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_6.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_7.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_8.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV045_sit_9.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV049_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV053_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV053_sit_2.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_10.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_2.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_3.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_4.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_5.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_6.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_7.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_8.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV057_sit_9.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV061_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV073_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV076_sit_1.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV078_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV078_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV081_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV081_sit_2.mp4 dosyasından 2 kare çekildi.


YOUTUBE_YouTubeCCTV084_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV086_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV086_sit_2.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV086_sit_3.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV086_sit_4.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV091_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV091_sit_2.mp4 dosyasından 5 kare çekildi.


YOUTUBE_YouTubeCCTV096_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV097_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV097_sit_2.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV097_sit_3.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV098_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV099_sit_1.mp4 dosyasından 1 kare çekildi.


YOUTUBE_YouTubeCCTV100_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV100_sit_2.mp4 dosyasından 4 kare çekildi.


YOUTUBE_YouTubeCCTV100_sit_3.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV101_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV102_sit_1.mp4 dosyasından 3 kare çekildi.


YOUTUBE_YouTubeCCTV102_sit_2.mp4 dosyasından 3 kare çekildi.
Toplam 8111 frame yüklendi.
Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_layer (InputLayer)    [(None, 128, 128, 3)]     0         
                                                                 
 conv2d (Conv2D)             (None, 128, 128, 64)      1792      
                                                                 
 batch_normalization (Batch  (None, 128, 128, 64)      256       
 Normalization)                                                  
                                                                 
 max_pooling2d (MaxPooling2  (None, 64, 64, 64)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 64, 64, 32)        18464     
                                      

In [ ]:
#Augmentation

import torchvision.transforms as T

frame_transform = T.Compose([
    T.RandomHorizontalFlip(p=0.5),
    T.RandomRotation(degrees=(-10, 10)),
    T.ColorJitter(brightness=0.3, contrast=0.3, saturation=0.2, hue=0.1),
    T.RandomApply([T.GaussianBlur(3)], p=0.3),
    T.RandomApply([T.RandomGrayscale(p=0.2)], p=0.2),
])



In [ ]:
import torch

def frame_dropout(frames, p=0.2):
    """Belirli bir ihtimalle rastgele kareleri düşür."""
    keep_mask = torch.rand(len(frames)) > p
    return frames[keep_mask] if keep_mask.sum() > 0 else frames

def speed_perturbation(frames, min_speed=0.8, max_speed=1.2):
    """Videoyu rastgele hızlandır veya yavaşlat."""
    speed_factor = torch.FloatTensor(1).uniform_(min_speed, max_speed).item()
    indices = torch.linspace(0, len(frames)-1, int(len(frames) * speed_factor)).long()
    return frames[indices]

In [ ]:
#Fixes frame count after applying frame dropout at augmentation

def fix_frame_count(frames, target_num_frames):
    """Frame sayısını sabitler: Eğer fazla ise kırpar, eksikse tekrarlar."""
    num_frames = frames.shape[0]  # Frame sayısını al

    if num_frames > target_num_frames:
        frames = frames[:target_num_frames]  # Fazla olanları kırp
    elif num_frames < target_num_frames:
        repeat_frames = target_num_frames - num_frames
        last_frame = frames[-1].unsqueeze(0).repeat(repeat_frames, 1, 1, 1)  # Eksik frame'leri tekrar et
        frames = torch.cat([frames, last_frame], dim=0)

    return frames  # Her durumda (T, C, H, W) döner

In [ ]:
import torch
import torchvision
import torch.nn.functional as F
from torch.utils.data import Dataset
import pandas as pd
import random

class VideoDataset(Dataset):
    def __init__(self, df, optical_flow_csv, num_frames=8, transform=None, resize_size=(192, 192), mode= "train"):
        self.df = df
        self.optical_flow_data = pd.read_csv(optical_flow_csv)  # 📌 Optical Flow verileri
        self.num_frames = num_frames
        self.transform = transform
        self.resize_size = resize_size
        self.mode = mode

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        video_path = self.df.iloc[idx]["video_path"]
        binary_label = self.df.iloc[idx]["binary_label"]
        multi_label = self.df.iloc[idx]["multi_label"]

        if binary_label == 1:  # **Anormal Video**
            # 📌 **Önceden hesaplanan en hareketli segmentleri al**
            segments = eval(self.optical_flow_data[self.optical_flow_data["video_path"] == video_path]["top_segments"].values[0])

            # 📌 **Rastgele bir hareketli segment seç**
            chosen_segment = random.choice(segments)

            # 📌 **Seçili segmentin frame'lerini al**
            frames = self.extract_frames(video_path, chosen_segment)

        else:  # **Normal Video**
            frames = self.extract_uniform_frames(video_path, self.num_frames)

        frames = fix_frame_count(frames, self.num_frames)

        #Eğer train setindeysek augmentation uygula
        if self.mode == "train":
            frames = frame_dropout(frames, p=0.1)
            frames = speed_perturbation(frames, min_speed=0.8, max_speed=1.2)
            frames = torch.stack([frame_transform(frame) for frame in frames])
            frames = fix_frame_count(frames,self.num_frames)


        # **🎯 GİRİŞ FORMATINI DÜZELTME**
        frames = frames.permute(1, 0, 2, 3).contiguous()  # (T, C, H, W) → (C, T, H, W)

        # Eğer transform varsa uygula
        if self.transform:
            frames = self.transform(frames)

        return frames, torch.tensor(binary_label, dtype=torch.float32), torch.tensor(multi_label, dtype=torch.long)

    def extract_frames(self, video_path, segment_index):
        """Anormal videolar için Optical Flow'a göre seçilen segmentten frame'leri çıkarır."""
        vr = torchvision.io.VideoReader(video_path, "video")
        vr.set_current_stream("video")

        frames = []
        frame_count = 0
        start_frame = segment_index  # 📌 Seçilen hareketli segmentin başlangıcı
        end_frame = start_frame + self.num_frames

        for i, frame in enumerate(vr):
            if i < start_frame:
                continue  # 📌 Seçilen segmentin başlangıcına kadar frame'leri atla

            if i >= end_frame:
                break  # 📌 Seçilen segment tamamlandıktan sonra çık

            frame = frame['data'].float() / 255.0  # Normalize (C, H, W)
            frame = F.interpolate(frame.unsqueeze(0), size=self.resize_size, mode="bilinear", align_corners=False)
            frames.append(frame.squeeze(0))

            frame_count += 1

        return torch.stack(frames)  # (T, C, H, W) olarak döndür

    def extract_uniform_frames(self, video_path, num_frames):
        """Normal videolar için eşit aralıklarla frame seçer."""
        vr = torchvision.io.VideoReader(video_path, "video")
        vr.set_current_stream("video")

        frames = []
        for frame in vr:
            frame = frame['data'].float() / 255.0  # Normalize (C, H, W)
            frame = F.interpolate(frame.unsqueeze(0), size=self.resize_size, mode="bilinear", align_corners=False)
            frames.append(frame.squeeze(0))

        total_frames = len(frames)

        # Eğer video kısa ise, son frame ile doldur
        if total_frames < num_frames:
            while len(frames) < num_frames:
                frames.append(frames[-1].clone())
        else:
            # 📌 **Frame'leri eşit aralıklarla seç**
            index_list = torch.linspace(0, total_frames - 1, num_frames).long()
            frames = [frames[i] for i in index_list]

        return torch.stack(frames)  # (T, C, H, W)

In [ ]:
train_dataset = VideoDataset(train_df, "/content/drive/MyDrive/modelepochsaves/optical_flow_segments.csv", num_frames=48, mode="train")
test_dataset = VideoDataset(test_df, "/content/drive/MyDrive/modelepochsaves/optical_flow_segments.csv", num_frames=48, mode="test")

train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=8, shuffle=False)

print(f"Train loader {len(train_loader)} batches")
print(f"Test loader {len(test_loader)} batches")

Train loader 326 batches
Test loader 82 batches


X3D modelimizin output layerini tüm classlarımızı kapsayacak şekilde arttırmamız lazım
Son katman ise ResNetBasicHead altında proj olarak tanımlanmış.

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision

# ✅ X3D Modelini Yükleme
model_name = "x3d_s"
base_model = torch.hub.load('facebookresearch/pytorchvideo', model_name, pretrained=True)

# ✅ **Çıkış Katmanlarını Güncelle**
in_features = base_model.blocks[-1].proj.out_features  # ✅ **Doğru çıkışı al!** (400 olmalı)

binary_head = nn.Linear(2048, 1)  # Binary sınıflandırma için Sigmoid
multi_head = nn.Linear(2048, len(train_df["multi_label"].unique()))  # Multi-class sınıflandırma

# ✅ **MultiTaskX3D Modeli**
class MultiTaskX3D(nn.Module):
    def __init__(self, base_model, binary_head, multi_head):
        super(MultiTaskX3D, self).__init__()

        # **Feature Extractor - Son bloğun içindeki ProjectedPool eklendi!**
        self.feature_extractor = nn.Sequential(
            *base_model.blocks[:-1],
            base_model.blocks[-1].pool  # ✅ **Burada 400'e dönüşüm sağlandı!**
        )

        # **Global Average Pooling - Tüm boyutları 1x1x1'e düşürüyor!**
        self.gap = nn.AdaptiveAvgPool3d(1)

        # **Flatten**
        self.flatten = nn.Flatten()

        # **Çıkış Katmanları**
        self.binary_head = binary_head
        self.binary_activation = nn.Sigmoid()

        self.multi_head = multi_head
        self.multi_activation = nn.Softmax(dim=1)

    def forward(self, x):
        #print(f"🚀 Model Girişi: {x.shape}")

        x = self.feature_extractor(x)
        #print(f"🟡 Feature Extractor Çıkışı: {x.shape}")  # **(16, 400, 1, 1, 1) olmalı!**

        x = self.gap(x)  # ✅ **GAP ile (16, 400, 1, 1, 1)**
        #print(f"🔵 GAP Sonrası Boyut: {x.shape}")

        x = self.flatten(x)  # ✅ **Artık (16, 400) olması lazım!**
        #print(f"🟢 Flatten Sonrası Boyut: {x.shape}")

        binary_out = self.binary_head(x)  # Binary sınıflandırma
        multi_out = self.multi_activation(self.multi_head(x))  # Multi-class sınıflandırma

        return binary_out, multi_out

# ✅ Modeli başlat
model = MultiTaskX3D(base_model, binary_head, multi_head)

device = "cuda" if torch.cuda.is_available() else "cpu"
model = model.to(device)

print(f"🚀 Model başarıyla yüklendi! Çalıştırılan cihaz: {device}")

Downloading: "https://github.com/facebookresearch/pytorchvideo/zipball/main" to /root/.cache/torch/hub/main.zip
Downloading: "https://dl.fbaipublicfiles.com/pytorchvideo/model_zoo/kinetics/X3D_S.pyth" to /root/.cache/torch/hub/checkpoints/X3D_S.pyth
100%|██████████| 29.4M/29.4M [00:00<00:00, 38.7MB/s]


🚀 Model başarıyla yüklendi! Çalıştırılan cihaz: cuda


import torch

model = model.to("cpu")
dummy_input = torch.randn(1, 3, 16, 160, 160)

torch.onnx.export(model, dummy_input, "x3d_xs.onnx", opset_version=11
                  ,do_constant_folding=True, input_names=["input"], output_names=["output"])

In [ ]:
from torch.nn import CrossEntropyLoss

class FocalLoss(nn.Module):
    def __init__(self, alpha=0.25, gamma=2.0):
        super(FocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma

    def forward(self,inputs,targets):
        ce_loss = F.cross_entropy(inputs,targets,reduction='none')
        pt = torch.exp(-ce_loss)
        focal_loss = self.alpha * (1-pt)**self.gamma * ce_loss
        return focal_loss.mean()

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision

epoch_file_location = "/content/drive/MyDrive/modelepochsaves"

# 🎯 Loss & Optimizer
criterion_binary = nn.BCEWithLogitsLoss()  # Binary classification için
criterion_multi = FocalLoss()  # Multi-class classification için

optimizer = optim.Adam(model.parameters(), lr=0.0001, weight_decay=1e-2)  # Weight decay ile overfitting azaltılır
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10)

# ⚡ AMP (Automatic Mixed Precision)
scaler = torch.amp.GradScaler()

# 🛑 Early Stopping Parametreleri
patience = 3  # Kaç epoch boyunca iyileşme olmazsa durdursun
min_delta = 0.01  # İyileşme için gereken minimum fark
best_loss = float("inf")  # En iyi validation loss başta sonsuz olarak ayarlanır
counter = 0  # İyileşme olmayan epoch sayısı
accuracy_threshold = 85.0  # Accuracy'nin geçmesi gereken eşik

device = "cuda" if torch.cuda.is_available() else "cpu"

for epoch in range(40):
    model.train()
    running_loss = 0.0

    for videos, binary_labels, multi_labels in train_loader:
        videos, binary_labels, multi_labels = (
            videos.to(device),
            binary_labels.to(device).float(),
            multi_labels.to(device),
        )

        optimizer.zero_grad()

        with torch.amp.autocast(device):
            binary_out, multi_out = model(videos)

            # 🟢 Binary Classification Loss (Sigmoid)
            binary_loss = criterion_binary(binary_out.squeeze(), binary_labels)

            # 🔴 Multi-Class Classification Loss (Softmax) → Sadece anormal olanlara uygula
            mask = (binary_labels == 1)
            if mask.sum() > 0:
                multi_loss = criterion_multi(multi_out[mask], multi_labels[mask])
            else:
                multi_loss = torch.tensor(0.0, device=device)

            # ✅ Toplam loss hesapla (Binary + Multi)
            total_loss = binary_loss + multi_loss

        scaler.scale(total_loss).backward()
        scaler.step(optimizer)
        scaler.update()

        running_loss += total_loss.item()

    train_loss = running_loss / len(train_loader)

    # 📌 VALIDATION AŞAMASI
    model.eval()
    val_loss = 0.0
    correct_binary = 0
    correct_multi = 0
    total_binary = 0
    total_multi = 0

    with torch.no_grad():
        for videos, binary_labels, multi_labels in test_loader:
            videos, binary_labels, multi_labels = (
                videos.to(device),
                binary_labels.to(device).float(),
                multi_labels.to(device),
            )

            binary_out, multi_out = model(videos)

            # ✅ Binary Accuracy (Normal mi, Anormal mi?)
            pred_binary = (binary_out.squeeze() > 0.5).long()
            correct_binary += (pred_binary == binary_labels).sum().item()
            total_binary += binary_labels.size(0)

            # ✅ Multi-Class Accuracy (Eğer anormalse hangi sınıf?)
            mask = (binary_labels == 1)
            if mask.sum() > 0:
                pred_multi = torch.argmax(multi_out[mask], dim=1)
                correct_multi += (pred_multi == multi_labels[mask]).sum().item()
                total_multi += multi_labels[mask].size(0)

            # ✅ Validation loss hesapla
            loss = criterion_binary(binary_out.squeeze(), binary_labels)
            if total_multi > 0:
                loss += criterion_multi(multi_out[mask], multi_labels[mask])
            val_loss += loss.item()

    val_loss /= len(test_loader)
    binary_acc = 100 * correct_binary / total_binary
    multi_acc = 100 * correct_multi / total_multi if total_multi > 0 else 0

    print(
        f"Epoch {epoch+1} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f} "
        f"| Binary Acc: {binary_acc:.2f}% | Multi Acc: {multi_acc:.2f}%"
    )

    # ✅ **Early Stopping (Binary Accuracy 85%'i geçmeli)**
    if binary_acc >= accuracy_threshold:
        if best_loss - val_loss > min_delta:
            best_loss = val_loss
            counter = 0
            torch.save(model.state_dict(), f"{epoch_file_location}/best_model_epoch_{epoch+1}.pth")
            print(f"✅ Model kaydedildi: best_model_epoch_{epoch+1}.pth")
        else:
            counter += 1
            print(f"❗ Early stopping counter: {counter}/{patience}")

        if counter >= patience:
            print("⏹️ Early stopping triggered. Training stopped.")
            break
    else:
        print(f"🚨 Accuracy {binary_acc:.2f}% < {accuracy_threshold}%, early stopping devre dışı!")

    scheduler.step()

    if (epoch + 1) % 5 == 0:
        torch.save(
            {
                "epoch": epoch,
                "model_state_dict": model.state_dict(),
                "optimizer_state_dict": optimizer.state_dict(),
            },
            f"{epoch_file_location}/checkpoint_{epoch}.pth",
        )
        print(f"Model {epoch + 1} saved to checkpoint_{epoch + 1}.pth")

print("✅ Eğitim tamamlandı!")

Epoch 1 | Train Loss: 0.7108 | Val Loss: 0.7181 | Binary Acc: 60.06% | Multi Acc: 48.96%
🚨 Accuracy 60.06% < 85.0%, early stopping devre dışı!
Epoch 2 | Train Loss: 0.6038 | Val Loss: 0.7167 | Binary Acc: 56.07% | Multi Acc: 53.11%
🚨 Accuracy 56.07% < 85.0%, early stopping devre dışı!
Epoch 3 | Train Loss: 0.5089 | Val Loss: 0.7046 | Binary Acc: 54.99% | Multi Acc: 54.77%
🚨 Accuracy 54.99% < 85.0%, early stopping devre dışı!
